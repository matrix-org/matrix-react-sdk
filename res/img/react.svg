<svg xmlns="http://www.w3.org/2000/svg" width="18" height="17" viewBox="0 0 18 17">
    <g fill="none" fill-rule="evenodd">
        <path fill="#2E2F32" fill-rule="nonzero" d="M13.6 14.5a7.875 7.875 0 1 1 2.183-3h-.818a7.125 7.125 0 1 0-2.62 3H13.6z"/>
        <g stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75">
            <path d="M11.5 13h6M14.5 10v6"/>
        </g>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75" d="M5.5 10s1.125 1.5 3 1.5 3-1.5 3-1.5"/>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M6.25 6.25h.008M10.75 6.25h.008"/>
    </g>
</svg>
