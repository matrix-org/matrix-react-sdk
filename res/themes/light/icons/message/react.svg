<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 464c-119.1 0-216-96.9-216-216S128.9 40 248 40s216 96.9 216 216-96.9 216-216 216zm90.2-146.2C315.8 352.6 282.9 368 248 368s-67.8-15.4-90.2-42.2c-5.7-6.8-15.8-7.7-22.5-2-6.8 5.7-7.7 15.7-2 22.5C161.7 380.4 203.6 400 248 400s86.3-19.6 114.8-53.8c5.7-6.8 4.8-16.9-2-22.5-6.8-5.6-16.9-4.7-22.6 2.1zM168 240c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32zm160 0c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32z"/></svg>