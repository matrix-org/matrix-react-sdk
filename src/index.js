/*
Copyright 2015, 2016 OpenMarket Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

var Skinner = require('./Skinner');

var t = require('counterpart');

t.registerTranslations('en', require('../languages/en'));
t.registerTranslations('pt-BR', require('../languages/pt-BR'));

t.setFallbackLocale('en');
t.setLocale(window.navigator.languages[0]);



module.exports.loadSkin = function(skinObject) {
    Skinner.load(skinObject);
};

module.exports.resetSkin = function() {
    Skinner.reset();
};

module.exports.getComponent = function(componentName) {
    return Skinner.getComponent(componentName);
};
