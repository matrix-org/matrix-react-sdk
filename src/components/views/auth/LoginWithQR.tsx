/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import { MSC3906Rendezvous, MSC3906RendezvousPayload, RendezvousFailureReason } from 'matrix-js-sdk/src/rendezvous';
import { MSC3886SimpleHttpRendezvousTransport } from 'matrix-js-sdk/src/rendezvous/transports';
import { MSC3903ECDHPayload, MSC3903ECDHv1RendezvousChannel } from 'matrix-js-sdk/src/rendezvous/channels';
import { logger } from 'matrix-js-sdk/src/logger';
import { MatrixClient } from 'matrix-js-sdk/src/client';

import { _t } from "../../../languageHandler";
import AccessibleButton from '../elements/AccessibleButton';
import QRCode from '../elements/QRCode';
import Spinner from '../elements/Spinner';
import { Icon as BackButtonIcon } from "../../../../res/img/element-icons/back.svg";
import { Icon as DevicesIcon } from "../../../../res/img/element-icons/devices.svg";
import { Icon as WarningBadge } from "../../../../res/img/element-icons/warning-badge.svg";
import { Icon as InfoIcon } from "../../../../res/img/element-icons/i.svg";
import { wrapRequestWithDialog } from '../../../utils/UserInteractiveAuth';

/**
 * The intention of this enum is to have a mode that scans a QR code instead of generating one.
 */
export enum Mode {
    /**
     * A QR code with be generated and shown
     */
    Show = "show",
}

enum Phase {
    Loading,
    ShowingQR,
    Connecting,
    Connected,
    WaitingForDevice,
    Verifying,
    Error,
}

interface IProps {
    client: MatrixClient;
    mode: Mode;
    onFinished(...args: any): void;
}

interface IState {
    phase: Phase;
    rendezvous?: MSC3906Rendezvous;
    confirmationDigits?: string;
    failureReason?: RendezvousFailureReason;
    mediaPermissionError?: boolean;
}

/**
 * A component that allows sign in and E2EE set up with a QR code.
 *
 * It implements both `login.start` and `login-reciprocate` capabilities as well as both scanning and showing QR codes.
 *
 * This uses the unstable feature of MSC3906: https://github.com/matrix-org/matrix-spec-proposals/pull/3906
 */
export default class LoginWithQR extends React.Component<IProps, IState> {
    public constructor(props) {
        super(props);

        this.state = {
            phase: Phase.Loading,
        };
    }

    public componentDidMount(): void {
        this.updateMode(this.props.mode).then(() => {});
    }

    public componentDidUpdate(prevProps: Readonly<IProps>): void {
        if (prevProps.mode !== this.props.mode) {
            this.updateMode(this.props.mode).then(() => {});
        }
    }

    private async updateMode(mode: Mode) {
        this.setState({ phase: Phase.Loading });
        if (this.state.rendezvous) {
            this.state.rendezvous.onFailure = undefined;
            await this.state.rendezvous.cancel(RendezvousFailureReason.UserCancelled);
            this.setState({ rendezvous: undefined });
        }
        if (mode === Mode.Show) {
            await this.generateCode();
        }
    }

    public componentWillUnmount(): void {
        if (this.state.rendezvous) {
            // eslint-disable-next-line react/no-direct-mutation-state
            this.state.rendezvous.onFailure = undefined;
            // calling cancel will call close() as well to clean up the resources
            this.state.rendezvous.cancel(RendezvousFailureReason.UserCancelled).then(() => {});
        }
    }

    private approveLogin = async (): Promise<void> => {
        if (!this.state.rendezvous) {
            throw new Error('Rendezvous not found');
        }
        this.setState({ phase: Phase.Loading });

        try {
            logger.info("Requesting login token");

            const { login_token: loginToken } = await wrapRequestWithDialog(this.props.client.requestLoginToken, {
                matrixClient: this.props.client,
                title: _t("Sign in new device"),
            })();

            this.setState({ phase: Phase.WaitingForDevice });

            const newDeviceId = await this.state.rendezvous.approveLoginOnExistingDevice(loginToken);
            if (!newDeviceId) {
                // user denied
                return;
            }
            if (!this.props.client.crypto) {
                // no E2EE to set up
                this.props.onFinished(true);
                return;
            }
            await this.state.rendezvous.verifyNewDeviceOnExistingDevice();
            this.props.onFinished(true);
        } catch (e) {
            logger.error('Error whilst approving sign in', e);
            this.setState({ phase: Phase.Error, failureReason: RendezvousFailureReason.Unknown });
        }
    };

    private generateCode = async () => {
        let rendezvous: MSC3906Rendezvous;
        try {
            const transport = new MSC3886SimpleHttpRendezvousTransport<MSC3903ECDHPayload>({
                onFailure: this.onFailure,
                client: this.props.client,
            });

            const channel = new MSC3903ECDHv1RendezvousChannel<MSC3906RendezvousPayload>(
                transport, undefined, this.onFailure,
            );

            rendezvous = new MSC3906Rendezvous(channel, this.props.client, this.onFailure);

            await rendezvous.generateCode();
            this.setState({
                phase: Phase.ShowingQR,
                rendezvous,
                failureReason: undefined,
            });
        } catch (e) {
            logger.error('Error whilst generating QR code', e);
            this.setState({ phase: Phase.Error, failureReason: RendezvousFailureReason.HomeserverLacksSupport });
            return;
        }

        try {
            const confirmationDigits = await rendezvous.startAfterShowingCode();
            this.setState({ phase: Phase.Connected, confirmationDigits });
        } catch (e) {
            logger.error('Error whilst doing QR login', e);
            // only set to error phase if it hasn't already been set by onFailure or similar
            if (this.state.phase !== Phase.Error) {
                this.setState({ phase: Phase.Error, failureReason: RendezvousFailureReason.Unknown });
            }
        }
    };

    private onFailure = (reason: RendezvousFailureReason) => {
        logger.info(`Rendezvous failed: ${reason}`);
        this.setState({ phase: Phase.Error, failureReason: reason });
    };

    public reset() {
        this.setState({
            rendezvous: undefined,
            confirmationDigits: undefined,
            failureReason: undefined,
        });
    }

    private cancelClicked = async (e: React.FormEvent) => {
        e.preventDefault();
        await this.state.rendezvous?.cancel(RendezvousFailureReason.UserCancelled);
        this.reset();
        this.props.onFinished(false);
    };

    private declineClicked = async (e: React.FormEvent) => {
        e.preventDefault();
        await this.state.rendezvous?.declineLoginOnExistingDevice();
        this.reset();
        this.props.onFinished(false);
    };

    private tryAgainClicked = async (e: React.FormEvent) => {
        e.preventDefault();
        this.reset();
        await this.updateMode(this.props.mode);
    };

    private onBackClick = async () => {
        await this.state.rendezvous?.cancel(RendezvousFailureReason.UserCancelled);

        this.props.onFinished(false);
    };

    private cancelButton = () => <AccessibleButton
        kind="primary_outline"
        onClick={this.cancelClicked}
    >
        { _t("Cancel") }
    </AccessibleButton>;

    private simpleSpinner = (description?: string): JSX.Element => {
        return <div className="mx_LoginWithQR_spinner">
            <div>
                <Spinner />
                { description && <p>{ description }</p> }
            </div>
        </div>;
    };

    public render() {
        let title: string;
        let titleIcon: JSX.Element | undefined;
        let main: JSX.Element | undefined;
        let buttons: JSX.Element | undefined;
        let backButton = true;
        let cancellationMessage: string | undefined;
        let centreTitle = false;

        switch (this.state.phase) {
            case Phase.Error:
                switch (this.state.failureReason) {
                    case RendezvousFailureReason.Expired:
                        cancellationMessage = _t("The linking wasn't completed in the required time.");
                        break;
                    case RendezvousFailureReason.InvalidCode:
                        cancellationMessage = _t("The scanned code is invalid.");
                        break;
                    case RendezvousFailureReason.UnsupportedAlgorithm:
                        cancellationMessage = _t("Linking with this device is not supported.");
                        break;
                    case RendezvousFailureReason.UserDeclined:
                        cancellationMessage = _t("The request was declined on the other device.");
                        break;
                    case RendezvousFailureReason.OtherDeviceAlreadySignedIn:
                        cancellationMessage = _t("The other device is already signed in.");
                        break;
                    case RendezvousFailureReason.OtherDeviceNotSignedIn:
                        cancellationMessage = _t("The other device isn't signed in.");
                        break;
                    case RendezvousFailureReason.UserCancelled:
                        cancellationMessage = _t("The request was cancelled.");
                        break;
                    case RendezvousFailureReason.Unknown:
                        cancellationMessage = _t("An unexpected error occurred.");
                        break;
                    case RendezvousFailureReason.HomeserverLacksSupport:
                        cancellationMessage = _t("The homeserver doesn't support signing in another device.");
                        break;
                    default:
                        cancellationMessage = _t("The request was cancelled.");
                        break;
                }
                title = _t("Connection failed");
                centreTitle = true;
                titleIcon = <WarningBadge className="error" />;
                backButton = false;
                main = <p data-testid="cancellation-message">{ cancellationMessage }</p>;
                buttons = <>
                    <AccessibleButton
                        kind="primary"
                        onClick={this.tryAgainClicked}
                    >
                        { _t("Try again") }
                    </AccessibleButton>
                    { this.cancelButton() }
                </>;
                break;
            case Phase.Connected:
                title = _t("Devices connected");
                titleIcon = <DevicesIcon className="normal" />;
                backButton = false;
                main = <>
                    <p>{ _t("Check that the code below matches with your other device:") }</p>
                    <div className="mx_LoginWithQR_confirmationDigits">
                        { this.state.confirmationDigits }
                    </div>
                    <div className="mx_LoginWithQR_confirmationAlert">
                        <div>
                            <InfoIcon />
                        </div>
                        <div>{ _t("By approving access for this device, it will have full access to your account.") }</div>
                    </div>
                </>;

                buttons = <>
                    <AccessibleButton
                        data-testid="decline-login-button"
                        kind="primary_outline"
                        onClick={this.declineClicked}
                    >
                        { _t("Cancel") }
                    </AccessibleButton>
                    <AccessibleButton
                        data-testid="approve-login-button"
                        kind="primary"
                        onClick={this.approveLogin}
                    >
                        { _t("Approve") }
                    </AccessibleButton>
                </>;
                break;
            case Phase.ShowingQR:
                title =_t("Sign in with QR code");
                if (this.state.rendezvous) {
                    const code = <div className="mx_LoginWithQR_qrWrapper">
                        <QRCode data={[{ data: Buffer.from(this.state.rendezvous.code), mode: 'byte' }]} className="mx_QRCode" />
                    </div>;
                    main = <>
                        <p>{ _t("Scan the QR code below with your device that's signed out.") }</p>
                        <ol>
                            <li>{ _t("Start at the sign in screen") }</li>
                            <li>{ _t("Select 'Scan QR code'") }</li>
                            <li>{ _t("Review and approve the sign in") }</li>
                        </ol>
                        { code }
                    </>;
                } else {
                    main = this.simpleSpinner();
                    buttons = this.cancelButton();
                }
                break;
            case Phase.Loading:
                main = this.simpleSpinner();
                break;
            case Phase.Connecting:
                main = this.simpleSpinner(_t("Connecting..."));
                buttons = this.cancelButton();
                break;
            case Phase.WaitingForDevice:
                main = this.simpleSpinner(_t("Waiting for device to sign in"));
                buttons = this.cancelButton();
                break;
            case Phase.Verifying:
                title = _t("Success");
                centreTitle = true;
                main = this.simpleSpinner(_t("Completing set up of your new device"));
                break;
        }

        return (
            <div data-testid="login-with-qr" className="mx_LoginWithQR">
                <div className={centreTitle ? "mx_LoginWithQR_centreTitle" : ""}>
                    { backButton ?
                        <AccessibleButton
                            data-testid="back-button"
                            className="mx_LoginWithQR_BackButton"
                            onClick={this.onBackClick}
                            title="Back"
                        >
                            <BackButtonIcon />
                        </AccessibleButton>
                        : null }
                    <h1>{ titleIcon }{ title }</h1>
                </div>
                <div className="mx_LoginWithQR_main">
                    { main }
                </div>
                <div className="mx_LoginWithQR_buttons">
                    { buttons }
                </div>
            </div>
        );
    }
}
