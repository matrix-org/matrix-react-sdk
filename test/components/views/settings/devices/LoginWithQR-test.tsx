/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { mocked } from 'jest-mock';
import React from 'react';
import { MSC3886SimpleHttpRendezvousTransport } from 'matrix-js-sdk/src/rendezvous/transports';
import { MSC3906Rendezvous, RendezvousFailureReason } from 'matrix-js-sdk/src/rendezvous';

import LoginWithQR, { Mode } from '../../../../../src/components/views/auth/LoginWithQR';
import type { MatrixClient } from 'matrix-js-sdk/src/matrix';
import { flushPromisesWithFakeTimers } from '../../../../test-utils';

jest.useFakeTimers();

jest.mock('matrix-js-sdk/src/rendezvous');
jest.mock('matrix-js-sdk/src/rendezvous/transports');
jest.mock('matrix-js-sdk/src/rendezvous/channels');

function makeClient() {
    return mocked({
        getUser: jest.fn(),
        isGuest: jest.fn().mockReturnValue(false),
        isUserIgnored: jest.fn(),
        isCryptoEnabled: jest.fn(),
        getUserId: jest.fn(),
        on: jest.fn(),
        isSynapseAdministrator: jest.fn().mockResolvedValue(false),
        isRoomEncrypted: jest.fn().mockReturnValue(false),
        mxcUrlToHttp: jest.fn().mockReturnValue('mock-mxcUrlToHttp'),
        doesServerSupportUnstableFeature: jest.fn().mockReturnValue(true),
        removeListener: jest.fn(),
        requestLoginToken: jest.fn(),
        currentState: {
            on: jest.fn(),
        },
    } as unknown as MatrixClient);
}

describe('<LoginWithQR />', () => {
    const client = makeClient();
    const defaultProps = {
        mode: Mode.Show,
        onFinished: jest.fn(),
    };
    const mockConfirmationDigits = 'mock-confirmation-digits';
    const newDeviceId = 'new-device-id';

    const getComponent = (props: { client: MatrixClient, onFinished?: () => void }) =>
        (<LoginWithQR {...defaultProps} {...props} />);

    beforeEach(() => {
        jest.clearAllMocks();
        jest.spyOn(MSC3906Rendezvous.prototype, 'generateCode').mockRestore();
        jest.spyOn(MSC3906Rendezvous.prototype, 'cancel').mockResolvedValue();
        jest.spyOn(MSC3906Rendezvous.prototype, 'declineLoginOnExistingDevice').mockResolvedValue();
        jest.spyOn(MSC3906Rendezvous.prototype, 'startAfterShowingCode').mockResolvedValue(mockConfirmationDigits);
        jest.spyOn(MSC3906Rendezvous.prototype, 'approveLoginOnExistingDevice').mockResolvedValue(newDeviceId);
        client.requestLoginToken.mockResolvedValue({
            login_token: 'token',
            expires_in: 1000,
        });
        // @ts-ignore
        client.crypto = undefined;
    });

    it('no content in case of no support', async () => {
        // simulate no support
        jest.spyOn(MSC3906Rendezvous.prototype, 'generateCode').mockRejectedValue('');
        const { container } = render(getComponent({ client }));
        await waitFor(() => screen.getAllByTestId('cancellation-message').length === 1);
        expect(container).toMatchSnapshot();
    });

    it('renders spinner while generating code', async () => {
        const { container } = render(getComponent({ client }));
        expect(container).toMatchSnapshot();
    });

    it('cancels rendezvous after user goes back', async () => {
        const { getByTestId } = render(getComponent({ client }));
        const rendezvous = mocked(MSC3906Rendezvous).mock.instances[0];
        // @ts-ignore assign to private prop
        rendezvous.code = 'rendezvous-code';

        // flush generate code promise
        await flushPromisesWithFakeTimers();

        fireEvent.click(getByTestId('back-button'));

        // wait for cancel
        await flushPromisesWithFakeTimers();

        expect(rendezvous.cancel).toHaveBeenCalledWith(RendezvousFailureReason.UserCancelled);
    });

    it('displays qr code after it is created', async () => {
        const { container, getByText } = render(getComponent({ client }));
        const rendezvous = mocked(MSC3906Rendezvous).mock.instances[0];
        // @ts-ignore assign to private prop
        rendezvous.code = 'rendezvous-code';

        await flushPromisesWithFakeTimers();

        expect(rendezvous.generateCode).toHaveBeenCalled();
        expect(getByText('Sign in with QR code')).toBeTruthy();
        expect(container).toMatchSnapshot();
    });

    it('displays confirmation digits after connected to rendezvous', async () => {
        const { container, getByText } = render(getComponent({ client }));
        const rendezvous = mocked(MSC3906Rendezvous).mock.instances[0];
        // @ts-ignore assign to private prop
        rendezvous.code = 'rendezvous-code';

        // flush generate code promise
        await flushPromisesWithFakeTimers();
        // flush waiting for connection promise
        await flushPromisesWithFakeTimers();

        expect(container).toMatchSnapshot();
        expect(getByText(mockConfirmationDigits)).toBeTruthy();
    });

    it('displays unknown error if connection to rendezvous fails', async () => {
        const { container } = render(getComponent({ client }));
        expect(MSC3886SimpleHttpRendezvousTransport).toHaveBeenCalledWith({
            onFailure: expect.any(Function),
            client,
        });
        const rendezvous = mocked(MSC3906Rendezvous).mock.instances[0];
        // @ts-ignore assign to private prop
        rendezvous.code = 'rendezvous-code';
        mocked(rendezvous).startAfterShowingCode.mockRejectedValue('oups');

        // flush generate code promise
        await flushPromisesWithFakeTimers();
        // flush waiting for connection promise
        await flushPromisesWithFakeTimers();

        expect(container).toMatchSnapshot();
    });

    it('declines login', async () => {
        const { getByTestId } = render(getComponent({ client }));
        const rendezvous = mocked(MSC3906Rendezvous).mock.instances[0];
        // @ts-ignore assign to private prop
        rendezvous.code = 'rendezvous-code';

        // flush generate code promise
        await flushPromisesWithFakeTimers();
        // flush waiting for connection promise
        await flushPromisesWithFakeTimers();

        fireEvent.click(getByTestId('decline-login-button'));

        expect(rendezvous.declineLoginOnExistingDevice).toHaveBeenCalled();
    });

    it('displays error when approving login fails', async () => {
        const { container, getByTestId } = render(getComponent({ client }));
        const rendezvous = mocked(MSC3906Rendezvous).mock.instances[0];
        // @ts-ignore assign to private prop
        rendezvous.code = 'rendezvous-code';
        client.requestLoginToken.mockRejectedValue('oups');

        // flush generate code promise
        await flushPromisesWithFakeTimers();
        // flush waiting for connection promise
        await flushPromisesWithFakeTimers();

        fireEvent.click(getByTestId('approve-login-button'));

        expect(client.requestLoginToken).toHaveBeenCalled();
        // flush token request promise
        await flushPromisesWithFakeTimers();
        await flushPromisesWithFakeTimers();

        expect(container).toMatchSnapshot();
    });

    it('approves login and waits for new device', async () => {
        const { container, getByTestId, getByText } = render(getComponent({ client }));
        const rendezvous = mocked(MSC3906Rendezvous).mock.instances[0];
        // @ts-ignore assign to private prop
        rendezvous.code = 'rendezvous-code';

        // flush generate code promise
        await flushPromisesWithFakeTimers();
        // flush waiting for connection promise
        await flushPromisesWithFakeTimers();

        fireEvent.click(getByTestId('approve-login-button'));

        expect(client.requestLoginToken).toHaveBeenCalled();
        // flush token request promise
        await flushPromisesWithFakeTimers();
        await flushPromisesWithFakeTimers();

        expect(getByText('Waiting for device to sign in')).toBeTruthy();
        expect(container).toMatchSnapshot();
    });

    it('does not continue with verification when user denies login', async () => {
        const onFinished = jest.fn();
        const { getByTestId } = render(getComponent({ client, onFinished }));
        const rendezvous = mocked(MSC3906Rendezvous).mock.instances[0];
        // @ts-ignore assign to private prop
        rendezvous.code = 'rendezvous-code';
        // no device id returned => user denied
        mocked(rendezvous).approveLoginOnExistingDevice.mockReturnValue(undefined);

        // flush generate code promise
        await flushPromisesWithFakeTimers();
        // flush waiting for connection promise
        await flushPromisesWithFakeTimers();

        fireEvent.click(getByTestId('approve-login-button'));

        // flush token request promise
        await flushPromisesWithFakeTimers();
        await flushPromisesWithFakeTimers();

        expect(rendezvous.approveLoginOnExistingDevice).toHaveBeenCalled();

        await flushPromisesWithFakeTimers();
        expect(onFinished).not.toHaveBeenCalled();
        expect(rendezvous.verifyNewDeviceOnExistingDevice).not.toHaveBeenCalled();
    });

    it('waits for device approval on existing device and finishes when crypto is not setup', async () => {
        const { getByTestId } = render(getComponent({ client }));
        const rendezvous = mocked(MSC3906Rendezvous).mock.instances[0];
        // @ts-ignore assign to private prop
        rendezvous.code = 'rendezvous-code';

        // flush generate code promise
        await flushPromisesWithFakeTimers();
        // flush waiting for connection promise
        await flushPromisesWithFakeTimers();

        fireEvent.click(getByTestId('approve-login-button'));

        // flush token request promise
        await flushPromisesWithFakeTimers();
        await flushPromisesWithFakeTimers();

        expect(rendezvous.approveLoginOnExistingDevice).toHaveBeenCalled();
        await flushPromisesWithFakeTimers();
        expect(defaultProps.onFinished).toHaveBeenCalledWith(true);
        // didnt attempt verification
        expect(rendezvous.verifyNewDeviceOnExistingDevice).not.toHaveBeenCalled();
    });

    it('waits for device approval on existing device and verifies device', async () => {
        const { getByTestId } = render(getComponent({ client }));
        const rendezvous = mocked(MSC3906Rendezvous).mock.instances[0];
        // @ts-ignore assign to private prop
        rendezvous.code = 'rendezvous-code';
        // we just check for presence of crypto
        // pretend it is set up
        // @ts-ignore
        client.crypto = {};

        // flush generate code promise
        await flushPromisesWithFakeTimers();
        // flush waiting for connection promise
        await flushPromisesWithFakeTimers();

        fireEvent.click(getByTestId('approve-login-button'));

        // flush token request promise
        await flushPromisesWithFakeTimers();
        await flushPromisesWithFakeTimers();

        expect(rendezvous.approveLoginOnExistingDevice).toHaveBeenCalled();
        // flush login approval
        await flushPromisesWithFakeTimers();
        expect(rendezvous.verifyNewDeviceOnExistingDevice).toHaveBeenCalled();
        // flush verification
        await flushPromisesWithFakeTimers();
        expect(defaultProps.onFinished).toHaveBeenCalledWith(true);
    });
});
