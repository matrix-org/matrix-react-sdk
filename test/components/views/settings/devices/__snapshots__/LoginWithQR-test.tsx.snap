// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginWithQR /> approves login and waits for new device 1`] = `
<div>
  <div
    class="mx_LoginWithQR"
    data-testid="login-with-qr"
  >
    <div
      class=""
    >
      <div
        class="mx_AccessibleButton mx_LoginWithQR_BackButton"
        data-testid="back-button"
        role="button"
        tabindex="0"
        title="Back"
      >
        <div />
      </div>
      <h1 />
    </div>
    <div
      class="mx_LoginWithQR_main"
    >
      <div
        class="mx_LoginWithQR_spinner"
      >
        <div>
          <div
            class="mx_Spinner"
          >
            <div
              aria-label="Loading..."
              class="mx_Spinner_icon"
              data-testid="spinner"
              role="progressbar"
              style="width: 32px; height: 32px;"
            />
          </div>
          <p>
            Waiting for device to sign in
          </p>
        </div>
      </div>
    </div>
    <div
      class="mx_LoginWithQR_buttons"
    >
      <div
        class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary_outline"
        role="button"
        tabindex="0"
      >
        Cancel
      </div>
    </div>
  </div>
</div>
`;

exports[`<LoginWithQR /> displays confirmation digits after connected to rendezvous 1`] = `
<div>
  <div
    class="mx_LoginWithQR"
    data-testid="login-with-qr"
  >
    <div
      class=""
    >
      <h1>
        <div
          class="normal"
        />
        Devices connected
      </h1>
    </div>
    <div
      class="mx_LoginWithQR_main"
    >
      <p>
        Check that the code below matches with your other device:
      </p>
      <div
        class="mx_LoginWithQR_confirmationDigits"
      >
        mock-confirmation-digits
      </div>
      <div
        class="mx_LoginWithQR_confirmationAlert"
      >
        <div>
          <div />
        </div>
        <div>
          By approving access for this device, it will have full access to your account.
        </div>
      </div>
    </div>
    <div
      class="mx_LoginWithQR_buttons"
    >
      <div
        class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary_outline"
        data-testid="decline-login-button"
        role="button"
        tabindex="0"
      >
        Cancel
      </div>
      <div
        class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary"
        data-testid="approve-login-button"
        role="button"
        tabindex="0"
      >
        Approve
      </div>
    </div>
  </div>
</div>
`;

exports[`<LoginWithQR /> displays error when approving login fails 1`] = `
<div>
  <div
    class="mx_LoginWithQR"
    data-testid="login-with-qr"
  >
    <div
      class="mx_LoginWithQR_centreTitle"
    >
      <h1>
        <div
          class="error"
        />
        Connection failed
      </h1>
    </div>
    <div
      class="mx_LoginWithQR_main"
    >
      <p
        data-testid="cancellation-message"
      >
        An unexpected error occurred.
      </p>
    </div>
    <div
      class="mx_LoginWithQR_buttons"
    >
      <div
        class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary"
        role="button"
        tabindex="0"
      >
        Try again
      </div>
      <div
        class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary_outline"
        role="button"
        tabindex="0"
      >
        Cancel
      </div>
    </div>
  </div>
</div>
`;

exports[`<LoginWithQR /> displays qr code after it is created 1`] = `
<div>
  <div
    class="mx_LoginWithQR"
    data-testid="login-with-qr"
  >
    <div
      class=""
    >
      <div
        class="mx_AccessibleButton mx_LoginWithQR_BackButton"
        data-testid="back-button"
        role="button"
        tabindex="0"
        title="Back"
      >
        <div />
      </div>
      <h1>
        Sign in with QR code
      </h1>
    </div>
    <div
      class="mx_LoginWithQR_main"
    >
      <p>
        Scan the QR code below with your device that's signed out.
      </p>
      <ol>
        <li>
          Start at the sign in screen
        </li>
        <li>
          Select 'Scan QR code'
        </li>
        <li>
          Review and approve the sign in
        </li>
      </ol>
      <div
        class="mx_LoginWithQR_qrWrapper"
      >
        <div
          class="mx_QRCode mx_QRCode"
        >
          <div
            class="mx_Spinner"
          >
            <div
              aria-label="Loading..."
              class="mx_Spinner_icon"
              data-testid="spinner"
              role="progressbar"
              style="width: 32px; height: 32px;"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="mx_LoginWithQR_buttons"
    />
  </div>
</div>
`;

exports[`<LoginWithQR /> displays unknown error if connection to rendezvous fails 1`] = `
<div>
  <div
    class="mx_LoginWithQR"
    data-testid="login-with-qr"
  >
    <div
      class="mx_LoginWithQR_centreTitle"
    >
      <h1>
        <div
          class="error"
        />
        Connection failed
      </h1>
    </div>
    <div
      class="mx_LoginWithQR_main"
    >
      <p
        data-testid="cancellation-message"
      >
        An unexpected error occurred.
      </p>
    </div>
    <div
      class="mx_LoginWithQR_buttons"
    >
      <div
        class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary"
        role="button"
        tabindex="0"
      >
        Try again
      </div>
      <div
        class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary_outline"
        role="button"
        tabindex="0"
      >
        Cancel
      </div>
    </div>
  </div>
</div>
`;

exports[`<LoginWithQR /> no content in case of no support 1`] = `
<div>
  <div
    class="mx_LoginWithQR"
    data-testid="login-with-qr"
  >
    <div
      class="mx_LoginWithQR_centreTitle"
    >
      <h1>
        <div
          class="error"
        />
        Connection failed
      </h1>
    </div>
    <div
      class="mx_LoginWithQR_main"
    >
      <p
        data-testid="cancellation-message"
      >
        The homeserver doesn't support signing in another device.
      </p>
    </div>
    <div
      class="mx_LoginWithQR_buttons"
    >
      <div
        class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary"
        role="button"
        tabindex="0"
      >
        Try again
      </div>
      <div
        class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary_outline"
        role="button"
        tabindex="0"
      >
        Cancel
      </div>
    </div>
  </div>
</div>
`;

exports[`<LoginWithQR /> renders spinner while generating code 1`] = `
<div>
  <div
    class="mx_LoginWithQR"
    data-testid="login-with-qr"
  >
    <div
      class=""
    >
      <div
        class="mx_AccessibleButton mx_LoginWithQR_BackButton"
        data-testid="back-button"
        role="button"
        tabindex="0"
        title="Back"
      >
        <div />
      </div>
      <h1 />
    </div>
    <div
      class="mx_LoginWithQR_main"
    >
      <div
        class="mx_LoginWithQR_spinner"
      >
        <div>
          <div
            class="mx_Spinner"
          >
            <div
              aria-label="Loading..."
              class="mx_Spinner_icon"
              data-testid="spinner"
              role="progressbar"
              style="width: 32px; height: 32px;"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="mx_LoginWithQR_buttons"
    />
  </div>
</div>
`;
